### PR TITLE
Add CRUD methods for Asset Attribute Values

### PR DESCRIPTION
**Pull request by greg5green**
## Why?
Need to be able to create, modify, find, and delete asset attribute values.

## What changed?
- Added methods for dealing with asset attribute values:
  - `AssetAttributes#createValue`
  - `AssetAttributes#deleteValue`
  - `AssetAttributes#getValuesByAssetId`
  - `AssetAttributes#getValuesByAttributeId`
  - `AssetAttributes#updateValue`

### CODE REVIEW

File: src/utils/assets/formatPaginatedDataFromServer.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  const _metadata = input._metadata || {};
+  const records = input.records || [];
+
+  return {
+    _metadata,
+    records: records.map(recordFormatter)
+  };
+}
+
+export default formatPaginatedDataFromServer;
```
Comment (leeallen337) [inline src/utils/assets/formatPaginatedDataFromServer.js]: Nice. I like this! Especially having the default formatter callback

File: src/assets/assetAttributes.js [hunk 2]
```diff
@@ -269,28 +290,264 @@ class AssetAttributes {
...
    * Creates a new asset attribute value
+   *
+   * API Endpoint: '/assets/:assetId/values'
+   * Method: POST
+   *
+   * @param {string} assetId The ID of the asset type (formatted as a UUID)
+   * @param {Object} assetAttributeValue
+   * @param {string} assetAttributeValue.assetAttributeId UUID corresponding to the asset attribute
+   * @param {string} assetAttributeValue.effectiveDate ISO 8601 Extended Format date/time string
+   * @param {string} assetAttributeValue.notes
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.js]: Yes, yes it is.

File: src/assets/assetAttributes.js [hunk 3]
```diff
@@ -269,28 +290,264 @@ class AssetAttributes {
...
+    }
+
+    return this._request.delete(
+      `${this._baseUrl}/assets/attributes/values/${assetAttributeValueId}`
+    );
+  }
 
   /**
-   * Gets an asset attribute value
+   * Gets the requested page of asset attribute values
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.js]: No page here -- changed to:
```
   * Gets asset attribute values for a particular asset
```

File: src/assets/assetAttributes.js [hunk 4]
```diff
@@ -269,28 +290,264 @@ class AssetAttributes {
...
+        params: formattedAssetAttributeFilters
+      })
+      .then((assetAttributeValues) =>
+        assetAttributeValues.map(formatAssetAttributeValueFromServer)
+      );
+  }
 
   /**
-   * Gets a list of all asset attribute values
+   * Gets the requested page of asset attribute values
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.js]: Changed to:
```
   * Gets a paginated list of asset attribute values for a particular attribute
   * of a particular asset
```

File: src/assets/assetAttributes.js [hunk 5]
```diff
@@ -269,28 +290,264 @@ class AssetAttributes {
...
+   * API Endpoint: '/assets/:assetId/attributes/values'
+   * Method: GET
+   *
+   * @param {String} assetId The ID of the asset for which you are looking up
+   *   attribute values  (formatted as a UUID)
+   * @param {Object} assetAttributeFilters Specific information that is used to
+   *   filter the list of asset attribute values
+   * @param {String} assetAttributeFilters.attributeLabel Label of the parent
+   *   asset attribute
+   * @param {String} assetAttributeFilters.effectiveDate Effective date of the
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.js]: Good catch.

File: src/assets/assetAttributes.js [hunk 6]
```diff
@@ -269,28 +290,264 @@ class AssetAttributes {
...
+   *
+   * @returns {Promise}
+   * @fulfill {AssetAttributeValue[]}
+   * @rejects {Error}
+   *
+   * @example
+   * contxtSdk.assets.attributes
+   *   .getValuesByAssetId(
+   *     'd7329ef3-ca63-4ad5-bb3e-632b702584f8'
+   *   )
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.js]: 👍

File: src/assets/assetAttributes.spec.js [hunk 7]
```diff
@@ -487,13 +486,487 @@ describe('Assets/Attributes', function() {
...
+      let valueFiltersToServerBeforeFormat;
+      let valuesFromServerAfterFormat;
+      let valuesFromServerBeforeFormat;
+
+      beforeEach(function() {
+        assetId = fixture.build('asset').id;
+        valueFiltersToServerAfterFormat = {
+          attributeLabel: faker.hacker.phrase(),
+          effectiveDate: faker.date.recent().toISOString()
+        };
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.spec.js]: 👍

File: src/assets/assetAttributes.spec.js [hunk 8]
```diff
@@ -487,13 +486,487 @@ describe('Assets/Attributes', function() {
...
+        valuesFromServerAfterFormat = fixture.buildList(
+          'assetAttributeValue',
+          faker.random.number({ min: 1, max: 20 }),
+          { assetId }
+        );
+        valuesFromServerBeforeFormat = valuesFromServerAfterFormat.map(
+          (values) =>
+            fixture.build('assetAttributeValue', values, { fromServer: true })
+        );
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.spec.js]: It doesn't really matter in the grand scheme, but it seems "more right" to make the objects be similar

File: src/assets/assetAttributes.spec.js [hunk 9]
```diff
@@ -487,13 +486,487 @@ describe('Assets/Attributes', function() {
...
+      });
+
+      it('formats the pagination options sent to the server', function() {
+        expect(formatPaginationOptionsToServer).to.be.calledWith(
+          paginationOptionsBeforeFormat
+        );
+      });
+
+      it('gets a list of asset attribute values from the server', function() {
+        expect(request.get).to.be.calledWith(
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.spec.js]: This is the author's stance on `.deep.calledWith`: https://github.com/domenic/sinon-chai/issues/109#issuecomment-322093025

File: src/utils/assets/formatAssetAttributeValueFromServer.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @param {Object} input
+ * @param {string} input.assetAttributeId UUID corresponding to the asset attribute
+ * @param {string} input.assetId UUID corresponding to the asset
+ * @param {string} input.createdAt ISO 8601 Extended Format date/time string
+ * @param {string} input.effectiveDate ISO 8601 Extended Format date/time string
+ * @param {string} input.id UUID
+ * @param {string} input.notes
+ * @param {string} input.updatedAt ISO 8601 Extended Format date/time string
+ * @param {string} input.value
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/utils/assets/formatAssetAttributeValueFromServer.js]: 1. Yes. I'll fix.
2. Yeah, probably. I keep forgetting if they should be optional or not in the formatters, but they should be.

File: src/utils/assets/formatAssetAttributeValueToServer.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ * Normalizes the asset attribute value object being sent to the API server
+ *
+ * @param {AssetAttributeValue} input
+ *
+ * @returns {Object} output
+ * @returns {string} output.assetAttributeId UUID corresponding to the asset attribute
+ * @returns {string} output.assetId UUID corresponding to the asset
+ * @returns {string} output.effectiveDate ISO 8601 Extended Format date/time string
+ * @returns {string} output.notes
+ * @returns {string} output.value
```
[the remark this answers is not in the document]
Comment (greg5green) [reply src/utils/assets/formatAssetAttributeValueToServer.js]: Fixing.

File: src/assets/assetAttributes.spec.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetAttributes.spec.js]: I guess I should have looked up `.deep.calledWith` before assuming it was a thing 😩

[the remark this answers is not in the document]
Comment (greg5green) [reply src/assets/assetAttributes.spec.js]: It really seems like it should be 😩

**Pull request merged**